### ISSUE TITLE
[Api request] totatlly disabling "Analog to Digital" for specific cores

### ISSUE THREAD

**Issue by barbudreadmon** [libretro/RetroArch#12844]
Follow-up of https://github.com/libretro/RetroArch/pull/12562

Some cores like FBNeo don't want that setting, because they are already dual mapping to dpad & analog left in a safe way, in which case i can only see that setting as dangerous (and i also heard some devices like the nintendo switch set it on by default). Having a mean to prevent that setting to be appliable at all for specific cores would be great.

**Comment by barbudreadmon** [comment]
tagging @jdgleaver about this

**Comment by hizzlekizzle** [comment]
Since AtoD mapping is a RetroArch thing rather than a libretro thing, I wouldn't think it would make sense to go into the API itself, but I was thinking RetroArch just recently got the ability to disable it on some cores automatically. I don't remember the circumstances that cause it to auto-disable, though. (I thought it was any time the analogs were actually initialized by the core, which should theoretically cover FBN, too, if it's doing it on its own)

**Comment by markwkidd** [comment]
This is an issue for MAME 2003/+ as well, where we also have analog to digital in the core.

**Comment by jdgleaver** [comment]
With #12562, AtoD mapping should always be disabled for cores that use analog input, unless the user explicitly chooses one of the 'forced' mapping options (which are never selected by default). So this issue report only covers these 'forced' options - i.e. should some cores override the user setting to guarantee properly working inputs?

> Since AtoD mapping is a RetroArch thing rather than a libretro thing, I wouldn't think it would make sense to go into the API itself

That's an excellent point! So if we did want to do something like this, it would probably be more appropriate to add a new field to the core info files - `adc_disabled` - to enable a frontend override without messing with the API.

**Comment by markwkidd** [comment]
@mahoneyt944 I wonder if you have thoughts and advice about this

**Comment by barbudreadmon** [comment]
One matter of concern is that, in the few topics about that setting, i always see users recommending to use the "forced" setting (there might be a reason the non-forced isn't popular ?).

>That's an excellent point! So if we did want to do something like this, it would probably be more appropriate to add a new field to the core info files - adc_disabled - to enable a frontend override without messing with the API.

It sounds like an excellent idea

**Comment by hizzlekizzle** [comment]
>So if we did want to do something like this, it would probably be more appropriate to add a new field to the core info files - adc_disabled - to enable a frontend override without messing with the API.

I had considered adding "needs_keyboard"/"needs_mouse" fields to the infos but didn't bother now that we can query the core directly (IIRC), but something like "needs_analogs = true" could auto-disable frontend AtoD.

On a related note, the info files now include a lot more information about the core capabilities if we wanted to auto-disable/hide some risky options, such as runahead and/or rewind (or savestates, or even 'controls' and/or 'cheats' menus). It could also be nice to include logging warnings based on these fields (e.g., "core needs fullpath, so softpatching is unavailable").

**Comment by mahoneyt944** [comment]
@markwkidd the forced options ignore analog input and send a digital signal only. The non-forced item only do this if the core does not support analog natively, therefore if it does the setting is ignored.

**Comment by grant2258** [comment]
That's not how we fixed it we had to map the digital and analog to directions. If you just map one only one will work for the relevant control

**Comment by mahoneyt944** [comment]
@grant2258 these are both a key part. Though mapping both is not enough for analog to work in a digital input game.

https://github.com/libretro/mame2003-plus-libretro/compare/Test

To test this theory, disabling our threshold check doesn't allow a analog stick to play mspacman even though they are both mapped....

We fixed the issue by mapping both inputs and by checking this threshold so that analog can be used in a digital game.

A to D simply makes pretends analog is a dpad

**Comment by grant2258** [comment]
It's always been this way @mahony just the refactoring caused some analog issues. It's what you map on the ui in plus.

**Comment by mahoneyt944** [comment]
> It's always been this way @mahony just the refactoring caused some analog issues. It's what you map on the ui in plus.. Goto input this game and just map the digital hat left analog controls won't work just the dhat mapping.
> 
> Sorry I had to edit it was on the phone.
> 
> Well I had a look never knew this option changed and had a look.
> 
> Analogue to digital type: left sends the analogue
> Analogue to digital type: left(forced) sends the digital dhat.
> 
> As far as I can tell this is how it should work perhaps renaming it to.
> Analogue to digital type: left (analogue)
> Analogue to digital type: left (digital)
> 
> and perhaps add the old behaviour back.
> Analogue to digital type: left (analogue & digital)
> 
> As far as i can see the issue has been resolved already if you understand what its doing. Before is was sending both signals.

Right that was my opinion as well. And maybe where some of the confusion is in talking about this 🙃.

The option Analog to Digital "left" send analog if the core natively supports analog or else it sends digital. The way I understand it is "left" basically is more decisive on what to send based on the cores hookup of analog. Forced just ignores everything and sends digital.

**Comment by grant2258** [comment]
How do your set the core to be digital only? This could be handy it would be even better it the retropad can tell you. Some controllers are monsters though. The dragonwise arcade controller defaults to analog but send 0 on off and full analog on press unless you press the mode button to put it digital then it sends to the dhat.

**Comment by mahoneyt944** [comment]
In 2003+ if you wanted true digital only I guess you'd have to unmap the analog inputs in the mame menu. We default to allowing both types for simplicity since this topic confuses a lot of users. They just want stuff to work.

I'm really not sure how a unique analog controller would react. I guess it could be tested?

**Comment by mahoneyt944** [comment]
Left (Forced) - is digital only

Left - is analog if supported by the core otherwise digital

**Comment by grant2258** [comment]
@mahoneyt944 "Left - is analog if supported by the core otherwise digital" Im not sure what you mean by this when is it not supported is it dependant on the retropad type?

@markwkidd   
Slightly off top looking at the input screen as far as the mouse goes looks great in udev you get this.

![udev](https://user-images.githubusercontent.com/6128601/130372987-0a086886-6036-490f-9c6d-511537ba6c07.png)

however when you get a driver that doesnt support device names you get this and it looks like the mouse isint available (linux x in this case)
![x](https://user-images.githubusercontent.com/6128601/130373015-6b752032-b8c9-4ef0-bd71-dc711d2ea8ca.png)
.

**Comment by mahoneyt944** [comment]
@grant2258  that's more of a question for @jdgleaver since I believe he configured how RA decides this or not. maybe he can elaborate on this. I assume it's looking to see if a call to an analog type was made or not?

**Comment by grant2258** [comment]
Im just curious myself would be a handy thing to know.

**Comment by mahoneyt944** [comment]
There's a few commits https://github.com/libretro/RetroArch/commit/3a82cf980dc9f1ce2a486614631ff8204b331ba1

You maybe able to find the answer to this.

**Comment by jdgleaver** [comment]
> @grant2258 that's more of a question for @jdgleaver since I believe he configured how RA decides this or not. maybe he can elaborate on this. I assume it's looking to see if a call to an analog type was made or not?

Please refer to the original comment for the PR that enabled this functionality: #12562

`disable mapping if core attempts to read analog input` means exactly that: it is disabled if a core polls analog stick input (i.e. it is purely dependent upon the core's behaviour)

**Comment by grant2258** [comment]
Makes sense if no analog is polled return dpad else act normal. I don't see any issues with this apart from mapping cores that don't use analog properly to use the l or r stick. This setting will have no issues for plus as we poll analog anyway.
